text/0567-use-and-resources.md

```diff
@@ -0,0 +1,953 @@
+}
+```
+
+```hbs
+<div
+  {{did-insert (perform this.loadProducts @searchUrl)}}
+  {{did-update (perform this.loadProducts @searchUrl)}}
+>
+</div>
+```
+
+We can see that this has a few disadvantages compared to the original approach:
+
+- It requires us to switch contexts between the template and the JS to
+  understand the flow of the component.
+- It requires us to explicitly list all of the dependencies for a task _twice_.
+  This is reminiscent of dependency lists in computed properties and observers,
+  and has similar drawbacks in some cases.
+- It requires us to add an element to our template, one which potentially does
+  nothing at all other than exist to allow us to invoke the modifiers.
+
+This approach does have one advantage over the original - it will only call
+perform if `@searchUrl` has actually changed. This would have potentially been a
+subtle bug in the original, since it could cause search results to reload if
+unrelated args changed. But the extra weight and other caveats make this
+solution _feel_ worse overall, as we've discovered through the experience of
+Octane's early adopters.
+
+### Introducing `@use` and Resources
+
+There are two core issues here:
+
+1. With Glimmer components, there is no simple way to define behavior that has
+   its own self contained lifecycle, independent of a template.
+2. Historically, even before Glimmer components, there has been no standard way
+   to mix imperative and declarative code, and expose the results naturally.
+   `ember-concurrency` comes close, but only solves one specific problem.
+
+The `@use` API aims to solve both of these issues, and to do so in a way that
+integrates with autotracking so that users don't have to list out dependencies
+constantly in a repetitive manner. The core of the API consists of:
+
+- The `@use` decorator (and the complementary `{{use}}` helper)
+- The `Resource` base class (implemented via a `ResourceManager`, similar to
+  both components and modifiers).
+
+With these, we can create the `RemoteData` example at the beginning of this RFC
+like so:
+
+```js
+export class RemoteData extends Resource {
+  @tracked data = null;
+  @tracked isLoading = false;
+
+  get state() {
+    return {
+      isLoading: this.isLoading,
+      data: this.data,
+    };
+  }
+
+  async start(url) {
+    this.isLoading = true;
+
+    let response = await fetch(url);
+    let data = await response.json();
+
+    if (!this.isDestroyed) {
+      this.isLoading = false;
+      this.data = data;
+    }
+  }
+}
+```
+
+And we would use it like so:
+
+```js
+export default class SearchResults extends Component {
+  @use products = RemoteData.create(this.args.searchUrl);
+}
+```
+
+```hbs
+{{#if this.products.isLoading}}
+  <LoadingSpinner/>
+{{else}}
+  {{#each this.products.data as |product|}}
+    <ProductCard @product={{product}} />
+  {{/each}}
+{{/if}}
+```
+
+### A Closer Look
+
+Resources fundamentally revolve around two lifecycle hooks: `start` and
+`teardown`.
```

Comment:
Yeah, that makes sense. I think `setup` and `teardown` could work, the main thing we want to avoid is implying that the hooks could occur on the same instance multiple times, since after `teardown` a new instance will always be created. I think `start` and `stop` do somewhat imply that (if it can be stopped, it can then started again, potentially), so `setup`/`teardown` work better IMO

Reply:
Updated in the latest draft to `setup`